itests/verifreg_test.go

```diff
@@ -503,3 +503,140 @@ func makeVerifier(ctx context.Context, t *testing.T, api *impl.FullNodeAPI, root
+
+	// make the 2 verifiers
+
+	makeVerifier(ctx, t, api, rootAddr, verifier1Addr)
+
+	// assign datacap to a client
+	initialDatacap := big.NewInt(20000)
+
+	params, err := actors.SerializeParams(&verifregst.AddVerifiedClientParams{Address: verifiedClientAddr, Allowance: initialDatacap})
+	require.NoError(t, err)
+
+	msg := &types.Message{
+		From:   verifier1Addr,
+		To:     verifreg.Address,
+		Method: verifreg.Methods.AddVerifiedClient,
+		Params: params,
+		Value:  big.Zero(),
+	}
+
+	sm, err := api.MpoolPushMessage(ctx, msg, nil)
+	require.NoError(t, err)
+
+	res, err := api.StateWaitMsg(ctx, sm.Cid(), 1, lapi.LookbackNoLimit, true)
+	require.NoError(t, err)
+	require.EqualValues(t, 0, res.Receipt.ExitCode)
+
+	// check datacap balance
+	dcap, err := api.StateVerifiedClientStatus(ctx, verifiedClientAddr, types.EmptyTSK)
+	require.NoError(t, err)
+	require.Equal(t, *dcap, initialDatacap)
+
+	minerId, err := address.IDFromAddress(miner.ActorAddr)
+	require.NoError(t, err)
+
+	allocationRequest1 := verifregst.AllocationRequest{
+		Provider:   abi.ActorID(minerId),
+		Data:       cid.MustParse("bafkqaaa"),
+		Size:       abi.PaddedPieceSize(initialDatacap.Uint64() / 2),
+		TermMin:    verifregst.MinimumVerifiedAllocationTerm,
+		TermMax:    verifregst.MinimumVerifiedAllocationTerm,
+		Expiration: verifregst.MaximumVerifiedAllocationExpiration,
+	}
+
+	allocationRequest2 := verifregst.AllocationRequest{
+		Provider:   abi.ActorID(minerId),
+		Data:       cid.MustParse("bafkqaaa"),
+		Size:       abi.PaddedPieceSize(initialDatacap.Uint64() / 2),
+		TermMin:    verifregst.MinimumVerifiedAllocationTerm,
+		TermMax:    verifregst.MinimumVerifiedAllocationTerm,
+		Expiration: verifregst.MaximumVerifiedAllocationExpiration,
+	}
+
+	allocationRequests := verifregst.AllocationRequests{
+		Allocations: []verifregst.AllocationRequest{allocationRequest1, allocationRequest2},
+	}
+
+	receiverParams, err := actors.SerializeParams(&allocationRequests)
+	require.NoError(t, err)
+
+	transferParams, err := actors.SerializeParams(&datacap2.TransferParams{
+		To:           builtin.VerifiedRegistryActorAddr,
+		Amount:       big.Mul(initialDatacap, builtin.TokenPrecision),
+		OperatorData: receiverParams,
+	})
+	require.NoError(t, err)
+
+	msg = &types.Message{
+		To:     builtin.DatacapActorAddr,
+		From:   verifiedClientAddr,
+		Method: datacap.Methods.TransferExported,
+		Params: transferParams,
+		Value:  big.Zero(),
+	}
+
+	sm, err = api.MpoolPushMessage(ctx, msg, nil)
+	require.NoError(t, err)
+
+	res, err = api.StateWaitMsg(ctx, sm.Cid(), 1, lapi.LookbackNoLimit, true)
+	require.NoError(t, err)
+	require.EqualValues(t, 0, res.Receipt.ExitCode)
+
+	// check datacap balance
+	dcap, err = api.StateVerifiedClientStatus(ctx, verifiedClientAddr, types.EmptyTSK)
+	require.NoError(t, err)
+	require.Nil(t, dcap)
+
+	allocations, err := api.StateGetAllAllocations(ctx, types.EmptyTSK)
+	require.NoError(t, err)
+
+	require.Equal(t, 2, len(allocations))
```

Comment:
would be good to check the returns here rather than just the length, maybe give them different piece CIDs and check those You could use `baga6ea4seaaqa` and `baga6ea4seaaqc` which are ~valid piece CIDs with different single byte digests.

There's also an itest in the actor events PR that'll introduce the whole allocation + claim lifecycle that we can piggy-back when we have both of these merged to check both allocations and claims at different stages.

Reply:
@rvagg Can you please provide me the link to the said PR? I can wait till it lands and create a more inclusive itest for this. Meanwhile, I will add the more detailed checks for the allocations.

Reply:
https://github.com/filecoin-project/lotus/pull/11618 is the latest iteration of it with a cleaner commit history (https://github.com/filecoin-project/lotus/pull/11540 has the gory details of the process)

It's much bigger in scope than this PR so I wouldn't want to hold this one up with that. You're welcome to rebase onto it to do the itest but I think it might just be best to notch this up as a TODO for then.

Reply:
I have added a TODO to the test. I can update it later once events PR is in place.